File: Solutions/Add Two Numbers/addTwoNumbers.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* addTwoNumbers(struct ListNode* l1, struct ListNode* l2) {
    struct ListNode*p = l1, *q = l2, *L, *temp;
    
    while (p && q) {     
        q->val = p->val = q->val + p->val;
        p = p->next;
        q = q->next;
    }
    
    L = (p) ? l1 : l2;  //l1和l2相比，哪个长度更长
    
    temp = L;
    while (temp) {
        if (temp->val >= 10) {
            if (!temp->next) temp->next = (struct ListNode*)calloc(1, sizeof(struct ListNode));
            temp->val -= 10;
            temp->next->val +=1;
        }
        temp = temp->next;
    }
    return L;
}

File: Solutions/445/addTwoNumbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
  
      // 单链表节点值存储在数组中
      let convertToArray = (l) => {
          let res = []
          while (l) {
              res.push(l.val)
              l = l.next
          }
  
          return res
      }
  
      // 两个数组值累加得出结果，再存入新单链表并返回
      let addTwoArray = (a1, a2) => {
          let dummy = new ListNode(0), node = null
          let len1 = a1.length -1, len2 = a2.length - 1, sum = 0
  
          while (len1 >= 0 || len2 >= 0 || sum) {
              if (len1 >= 0) sum += a1[len1--]
              if (len2 >= 0) sum += a2[len2--]
  
              node = new ListNode(sum % 10)
              sum = parseInt(sum / 10)
              node.next = dummy.next
              dummy.next = node
          }
          return dummy.next
      }
  
      return addTwoArray(convertToArray(l1), convertToArray(l2))
  
  };